File: apple/Murmur/History/HistoryViewModel.swift
import Foundation
import MurmurShared
import UIKit

/// The History screen's list, search and row actions.
///
/// Reads are synchronous on purpose: SQLite on the local file system answers a fifty-row query
/// in microseconds, and the alternative — an async reload per keystroke — buys nothing but a
/// chance for results to arrive out of order.
@MainActor
final class HistoryViewModel: ObservableObject {
    @Published var query = ""
    @Published var items: [Transcript] = []
    /// The row whose "Copied" check is showing, if any.
    @Published var copiedID: Int64?

    /// How many rows the screen keeps in memory. Search runs in SQLite, so this is a display
    /// cap, not a search cap.
    static let pageSize = 200

    private let history: HistoryStore
    private let pasteboard: (String) -> Void
    private var copyResetTask: Task<Void, Never>?

    init(
        history: HistoryStore,
        copy: @escaping (String) -> Void = { UIPasteboard.general.string = $0 }
    ) {
        self.history = history
        pasteboard = copy
    }

    /// Re-reads the list for the current ``query``. A database failure leaves the previous
    /// rows on screen rather than blanking the list under the user.
    func reload() {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let rows = try? history.list(limit: Self.pageSize, matching: trimmed.isEmpty ? nil : trimmed) else { return }
        items = rows
    }

    func delete(_ transcript: Transcript) {
        guard let id = transcript.id else { return }
        try? history.delete(id: id)
        reload()
    }

    /// Copies the cleaned text and shows a check on that row for a moment. The raw text is
    /// never what gets copied — the cleaned line is what the user asked Murmur for.
    func copy(_ transcript: Transcript) {
        pasteboard(transcript.cleanText)
        copiedID = transcript.id
        copyResetTask?.cancel()
        copyResetTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 1_500_000_000)
            guard !Task.isCancelled else { return }
            await MainActor.run {
                guard let self, self.copiedID == transcript.id else { return }
                self.copiedID = nil
            }
        }
    }
}

File: android/app/src/main/java/com/murmur/app/ui/history/HistoryViewModel.kt
package com.murmur.app.ui.history

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.murmur.app.data.history.HistoryStore
import com.murmur.app.data.history.TranscriptEntity

/**
 * The History screen's list, search and row actions. The iOS twin is `HistoryViewModel`.
 *
 * The search itself runs in SQLite rather than over the loaded rows, so [PAGE_SIZE] is a
 * display cap and not a search cap: a phrase from six months ago is still findable.
 */
class HistoryViewModel(private val history: HistoryStore) : ViewModel() {

    /** The search field. The screen reloads on a change; typing never blocks on SQLite. */
    var query by mutableStateOf("")

    var items by mutableStateOf<List<TranscriptEntity>>(emptyList())
        private set

    /** The row whose "Copied" check is showing, if any. */
    var copiedId by mutableStateOf<Long?>(null)

    /** The row whose raw transcript is expanded — the long-press "Show raw". */
    var rawShownId by mutableStateOf<Long?>(null)
        private set

    /** Newest first, filtered by [query] when it has anything in it. */
    suspend fun reload() {
        val trimmed = query.trim()
        items = history.list(limit = PAGE_SIZE, query = trimmed.ifEmpty { null })
    }

    suspend fun delete(id: Long) {
        history.delete(id)
        if (rawShownId == id) rawShownId = null
        reload()
    }

    fun toggleRaw(id: Long) {
        rawShownId = if (rawShownId == id) null else id
    }

    companion object {
        /** How many rows the screen keeps in memory. */
        const val PAGE_SIZE = 200
    }
}
